src/codegen/util/csv_scanner.cpp

```diff
@@ -0,0 +1,375 @@
+    // Check if we can even allocate any more bytes
+    if (static_cast<uint64_t>(len) > kMaxAllocSize - line_len_) {
+      const auto msg = StringUtil::Format(
+          "Line %u in file '%s' exceeds maximum line length: %lu",
+          line_number_ + 1, file_path_.c_str(), kMaxAllocSize);
+      throw Exception(msg);
+    }
+
+    // The current line buffer isn't large enough to store the new bytes, so we
+    // need to resize it. Let's find an allocation size large enough to fit the
+    // new bytes.
+    uint32_t new_maxlen = line_maxlen_ * 2;
+    while (new_maxlen < len) {
+      new_maxlen *= 2;
+    }
+
+    // Clamp
+    new_maxlen = std::min(new_maxlen, static_cast<uint32_t>(kMaxAllocSize));
+
+    auto *new_line = static_cast<char *>(memory_.Allocate(new_maxlen));
+
+    // Copy the old data
+    PELOTON_MEMCPY(new_line, line_, line_len_);
+
+    // Free old old
+    memory_.Free(line_);
+
+    // Setup pointers and sizes
+    line_ = new_line;
+    line_maxlen_ = new_maxlen - 1;
+
+    stats_.num_reallocs++;
+  }
+
+  // Copy provided data into the line buffer, ensuring null-byte termination.
+  PELOTON_MEMCPY(line_ + line_len_, data, len);
+  line_[line_len_ + len] = '\0';
+
+  // Increase the length of the line
+  line_len_ += len;
+
+  // Track copy stats
+  stats_.num_copies++;
+}
+
+// The objective of this function is to find a complete line in the CSV file.
+// The returned value will be a valid pointer to a null-terminated string that
+// is the next line in the CSV to be processed.
+char *CSVScanner::NextLine() {
+  line_len_ = 0;
+
+  const char quote = quote_;
+  const char escape = (quote_ == escape_ ? static_cast<char>('\0') : escape_);
+
+  bool in_quote = false;
+  bool last_was_escape = false;
+
+  const char *buf = buffer_;
+  uint32_t curr_buffer_pos = buffer_pos_;
+
+  while (true) {
+    if (curr_buffer_pos == buffer_end_) {
+      // We need to read more data from the CSV file. But first, we need to copy
+      // all the data in the read-buffer (i.e., [buffer_begin_, buffer_end_] to
+      // the line-buffer.
+      if (buffer_pos_ < curr_buffer_pos) {
+        AppendToLineBuffer(buffer_ + buffer_pos_,
+                           curr_buffer_pos - buffer_pos_);
+        buffer_pos_ = curr_buffer_pos;
+      }
+
+      // Reset positions
+      curr_buffer_pos = 0;
+
+      // Now, read more data
+      if (!NextBuffer()) {
+        // We hit en EOF
+        break;
+      }
+    }
+
+    // Read character
+    char c = buf[curr_buffer_pos++];
+
+    if (in_quote && c == escape) {
+      last_was_escape = !last_was_escape;
+    }
+    if (c == quote && !last_was_escape) {
+      in_quote = !in_quote;
+    }
+    if (c != escape) {
+      last_was_escape = false;
+    }
+
+    // Process the new-line character. If we a new-line and we're not currently
+    // in a quoted section, we're done.
+    if (c == '\n' && !in_quote) {
+      break;
+    }
+  }
+
+  // Flush remaining valid bytes
```

Comment:
I don't get this part, can you explain briefly? (When does this occur?)

Reply:
Essentially, all bytes between `buffer_pos_` and `curr_buffer_pos` are valid bytes that should be part of the current line. It is possible that we exit the loop not having written all valid bytes to the line buffer. This section ensures we flush these valid bytes before retuning.